cosmos/x/evm/plugins/txpool/mempool/mempool_reader.go

```diff
@@ -0,0 +1,187 @@
+// SPDX-License-Identifier: BUSL-1.1
+//
+// Copyright (C) 2023, Berachain Foundation. All rights reserved.
+// Use of this software is govered by the Business Source License included
+// in the LICENSE file of this repository and at www.mariadb.com/bsl11.
+//
+// ANY USE OF THE LICENSED WORK IN VIOLATION OF THIS LICENSE WILL AUTOMATICALLY
+// TERMINATE YOUR RIGHTS UNDER THIS LICENSE FOR THE CURRENT AND ALL OTHER
+// VERSIONS OF THE LICENSED WORK.
+//
+// THIS LICENSE DOES NOT GRANT YOU ANY RIGHT IN ANY TRADEMARK OR LOGO OF
+// LICENSOR OR ITS AFFILIATES (PROVIDED THAT YOU MAY USE A TRADEMARK OR LOGO OF
+// LICENSOR AS EXPRESSLY REQUIRED BY THIS LICENSE).
+//
+// TO THE EXTENT PERMITTED BY APPLICABLE LAW, THE LICENSED WORK IS PROVIDED ON
+// AN “AS IS” BASIS. LICENSOR HEREBY DISCLAIMS ALL WARRANTIES AND CONDITIONS,
+// EXPRESS OR IMPLIED, INCLUDING (WITHOUT LIMITATION) WARRANTIES OF
+// MERCHANTABILITY, FITNESS FOR A PARTICULAR PURPOSE, NON-INFRINGEMENT, AND
+// TITLE.
+
+package mempool
+
+import (
+	sdk "github.com/cosmos/cosmos-sdk/types"
+
+	cosmlib "pkg.berachain.dev/polaris/cosmos/lib"
+	evmtypes "pkg.berachain.dev/polaris/cosmos/x/evm/types"
+	"pkg.berachain.dev/polaris/eth/common"
+	coretypes "pkg.berachain.dev/polaris/eth/core/types"
+	"pkg.berachain.dev/polaris/lib/utils"
+)
+
+// Get is called when a transaction is retrieved from the mempool.
+func (etp *EthTxPool) Get(hash common.Hash) *coretypes.Transaction {
+	return etp.ethTxCache[hash]
+}
+
+// Pending is called when txs in the mempool are retrieved.
+func (etp *EthTxPool) Pending(bool) map[common.Address]coretypes.Transactions {
+	etp.mu.RLock()
+	defer etp.mu.RUnlock()
+
+	allNonces := etp.senderIndices
+	pending := make(map[common.Address]coretypes.Transactions)
+	for sender, list := range allNonces {
+		// get Eth Address of sender
+		addr := getEthAddressFromBech32(sender)
+
+		var pendingNonce int64 = -1
+
+		// add the first eth tx in the list, if it exists
+		var ethTx *coretypes.Transaction
+		for elem := list.Front(); elem != nil; elem = elem.Next() {
+			if ethTx = evmtypes.GetAsEthTx(utils.MustGetAs[sdk.Tx](elem.Value)); ethTx != nil {
+				switch {
+				case pendingNonce == -1:
+					// If its the first tx, set the pending nonce to the nonce of the tx.
+					pending[addr] = append(pending[addr], ethTx)
+					pendingNonce = int64(ethTx.Nonce())
+					// If on the first lookup the nonce delta is more than 0, then there is a gap and thus no pending transactions,
+					// but there are queued transactions. We continue
+					if sdbNonce := etp.nr.GetNonce(addr); uint64(pendingNonce)-sdbNonce >= 1 {
```

Comment:
if statedb nonce is 4, and the eth tx nonce is 5, that is valid right. so the tx with nonce 5 is pending. I think the case should be:
`uint64(pendingNonce) - sdbNonce > 1`

Reply:
👍

Reply:
no its invalid, the first txn to enter the mempool should == the statedb nonce.